### ISSUE TITLE
aggregate objects in single table - to support join.

### ISSUE THREAD

**Issue by JoshiManjila** [Impetus/kundera#762]
Sir,

I have one suggestion to overcome join problem. Instead of keeping an aggregate object like order with line-items in two different table spaces, keep them a single table space with list of line items as a List type. I think this approach will overcome the the join issue to a great extend.  It can be followed  with  any one-to-many relations. Cassandra way of designing table also suggest  to analyse  queries involved to decide on the tables  design.

Would you explain why it it  not tried in this way ?

Joshi

Joshi
.

**Comment by karthikprasad13** [comment]
Hi @JoshiManjila 

There is a difference between entity-relationships (one to many, one to one, etc) and Lists. While modeling the data, some users might want to `normalize` the data and use relation model and some are good with `denormalized` data (Like you suggested). Kundera being a JPA compliant ORM, should be able to solve both the use cases. That is one of the main reasons for this implementation.

In your case you can use `Element Collections`, which is a collection/map of objects.

Please refer this [Entity](https://github.com/impetus-opensource/Kundera/blob/4ab46233e39b87c5516e0f870dd581b79eaac265/src/kundera-cassandra/cassandra-core/src/test/java/com/impetus/client/cassandra/udt/PersonUDT.java) and [Test Case](https://github.com/impetus-opensource/Kundera/blob/4ab46233e39b87c5516e0f870dd581b79eaac265/src/kundera-cassandra/cassandra-core/src/test/java/com/impetus/client/cassandra/udt/UdtAsEmbeddablesCRUDTest.java)

-Karthik

**Comment by JoshiManjila** [comment]
Dear Sir,

Thank you for your quick response. If we are using demoralized form with element collection, then you are saying that both root and line-items are stored in a single key-space. Then can I   retrieve data from the childUDT in a single query. like

select \* from order where order.lineItem.name= "somename" .

Thanks

Joshi

**Comment by karthikprasad13** [comment]
Hi @JoshiManjila 

No. Since querying on fields inside UDT's is not yet supported by Cassandra (Check [CASSANDRA-6382](https://issues.apache.org/jira/browse/CASSANDRA-6382)) this is not possible as of now.

-Karthik.

**Comment by JoshiManjila** [comment]
Dear Karthik,

But is partially fixed and can able to create one index on UDTs from 2.1.3 right?

**Comment by chhavigangwal** [comment]
@JoshiManjila 

Yes, you can create indexes on the entire UDT field but not on its column specifically. 

Chhavi

**Comment by JoshiManjila** [comment]
Dear Chhavingwal,

Cassandra has a fix on udt index it in such a way that we can create one index on each column of UDT. But for collection objects they have  not  done yet. Also multiple index on same column also not supporting now . Though with fix they have done, it is enough for a one one mapping with @Embedded UDT. .We don't find it working with kundera at present version 2.17. 

Then  a query like 
select \* from patient._tble where emergecyContact.name="somename" 
would work. Here emergencyContact is a UDT that is embedded on Patient table.  

We really doubt probably it might be an implementation bug. Please share your thought on this.

Thanking you

Joshi

**Comment by karthikprasad13** [comment]
Hi @JoshiManjila 

> Cassandra has a fix on udt index it in such a way that we can create one index on each column of UDT.

The above feature is planned in 3.x release and there is an issue on JIRA regarding this (link in previous comments). If otherwise, can you share a reference to any documentation that you found?

-Karthik

**Comment by JoshiManjila** [comment]
Dear Karthik,

You might be true. Then how long we need to wait for Cassandra 3.0. or any  work around?

**Comment by karthikprasad13** [comment]
@JoshiManjila 

Dear Joshi,

There has been no official announcement of the release date of Cassandra 3.0.

Work arounds:
1) You can have client side logic for filtering the data. But considering this in big data domain will be very inefficient.
2) You can use any search technology (like [elasticsearch](https://www.elastic.co/products/elasticsearch)) for indexing your data and querying.

-Karthik
